[1216] Add content status

benilovj opened the pull request:
### Context
The course enrichment status is shown to publishers in the course table

### Changes proposed in this pull request
Add `content_status` to Course API V2.

### Guidance to review
This does a query per-course, which means N+1 queries on the courses table. We may need to consider performance going forward.

defong commented on app/models/course_enrichment.rb:
```diff
@@ -0,0 +1,23 @@
+#  created_by_user_id           :integer
+#  created_at                   :datetime         not null
+#  provider_code                :text             not null
+#  json_data                    :jsonb
```
Don't we need to map this back out?

benilovj replied:
We don’t need it in this PR.

defong commented:
Looking good, now why has travis gone on holiday...

heidar commented on app/models/course.rb:
```diff
@@ -104,4 +104,29 @@ def description
+    # `course_id` and `provider_id`. This should be fixed when the UCAS data model
+    # is revisited.
+    CourseEnrichment.where(
+      provider_code: self.provider.provider_code,
```
Redundant `self`

benilovj replied:
This whole method has been removed.

heidar commented on app/models/course.rb:
```diff
@@ -104,4 +104,29 @@ def description
+    # is revisited.
+    CourseEnrichment.where(
+      provider_code: self.provider.provider_code,
+      ucas_course_code: self.course_code
```
Redundant `self`

benilovj replied:
This whole method has been removed.

heidar commented on app/models/course.rb:
```diff
@@ -104,4 +104,29 @@ def description
+  end
+
+  def content_status
+    newest_enrichment = self.enrichments.order('created_at desc').first
```
Redundant `self`

benilovj replied:
fixed

heidar commented on spec/factory_specs/course_spec.rb:
```diff
@@ -1,18 +1,31 @@
-    expect(course).to be_instance_of(Course)
-    expect(course).to be_valid
-  end
+  it { should be_instance_of(Course) }
```
Do we really need to test for this?

misaka replied:
It may be pedantic, but I like having tests that prove the factories work. Also, it's CHEAP!

heidar commented on spec/models/course_enrichment_spec.rb:
```diff
@@ -0,0 +1,34 @@
+
+require 'rails_helper'
+
+describe CourseEnrichment, type: :model do
```
We've set up RSpec to infer types, not that I mind it being explicit.

benilovj replied:
All the other specs specify type so I'd prefer to be consistent and remove type wholesale on a separate PR, if we want to do that.

heidar commented on spec/models/course_enrichment_spec.rb:
```diff
@@ -0,0 +1,34 @@
+require 'rails_helper'
+
+describe CourseEnrichment, type: :model do
+  context 'initial draft' do
```
I feel like these contexts don't read very well. I think it would be clearer if it was "when the status is initial draft" or something like that?

benilovj replied:
fixed

heidar commented on spec/models/course_spec.rb:
```diff
@@ -333,4 +333,31 @@
       it { should_not include(course) }
     end
   end
```
I don't see a unit test for `#enrichments`, which is a public method.

benilovj replied:
added

heidar commented on app/models/course.rb:
```diff
@@ -55,6 +55,12 @@ class Course < ApplicationRecord
   has_many :site_statuses
   has_many :sites, through: :site_statuses
 
+  has_many :enrichments,
```
very nice

benilovj commented:
@heidar I've addressed all your comments. Ready for re-review.

@misaka I backed out your association additions to `CourseEnrichment` because they were causing the specs to fail (and anyway aren't needed as part of this change).
